core/app/models/spree/product.rb

```diff
@@ -161,6 +161,11 @@ def discontinued?
       !!discontinue_on && discontinue_on <= Time.current
     end
 
+    # determine if any variant (including master) can be supplied
+    def can_supply?
+      variants_including_master.any?(&:can_supply?)
```

Comment:
@priyank-gupta @Mafi88 even if master has some stock and all variants are sold out user can't add them to cart because javascript will disable the cart form's submit button